README.md
# php-beautiful-numbers

## 0. __construct ##

*php-beautiful-numbers* is a rather small number format class (< 15 kB with annotations). It handles numbers in a way that it outputs pretty, easy human readable numbers. 

When initialising you specify at least the tongue you want to use it in, and maybe further options like accuracy (see quick manual inside class file).  

```php
$bn = new bnformat\bnformat( ['lang'=>'de'] ); // set output to German 
```


## 1. sinum() – SI numbers ##

Not only in the physics department it is good practice to use the [SI format](https://en.wikipedia.org/wiki/International_System_of_Units) for writing down any number (in particular very large or small numbers). This ensures easy readability (thanks to the SI prefixes n, µ, m, k, M, G, ...) and only produces an output as precise as necessary (usually 3 digits are the sweet spot, meaning the total number of [significant figures](https://en.wikipedia.org/wiki/Significant_figures)). 

```php
echo $bn->sinum( 419532 ); // simple number without unit
echo $bn->sinum( 73672048.352987 );
echo $bn->sinum( 9.8437291615846E-5, 's' ); // with unit
echo $bn->sinum( 711372, 'B', ['bin'=>true] ); // use binary prefixes 
echo $bn->sinum( 3657.3480260881, 'm', ['acc'=>2] ); // accuracy = 2 digits 
```

Output looks like this (Deutsch, English):

```html
420 k  (= 419532) // korrekt gerundet, 3 Stellen Genauigkeit
73,7 M  (= 73672048.352987) // dito
98,4 µs  (= 9.8437291615846E-5 Sekunde) // mit Einheit und Prefix
695 KiB  (= 711372 Byte) // mit Binärprefix
3,7 km  (= 3657.3480260881 Meter) // Genauigkeit 2 Stellen
```
```html
420 k  (= 419532) // properly rounded, 3 digits accuracy
73.7 M  (= 73672048.352987) // same
98.4 µs  (= 9.8437291615846E-5 second) // with unit and prefix
695 KiB  (= 711372 byte) // using binary prefix
3.7 km  (= 3657.3480260881 meter) // accuracy 2 digits
```

If you use HTML mode each space before the unit/prefix get replaced by a [thin space](https://en.wikipedia.org/wiki/Thin_space). Also (if not disabled) any ambiguous [significant zero](https://en.wikipedia.org/wiki/Significant_figures#Identifying_significant_figures) automatically gets overlined to show the given precision.

## 2.1. tnum() – text numbers ##

In newspapers and other running text it is common practice to note the numbers from 0 to 12 written-out; all other numbers are written as digits. This produces more beautiful and easier to read texts. (Additionally this function automatically rounds to a given accuracy when you want to display large numbers.) 

```php
echo "I see " . $bn->tnum( $val ) . " trees on the hill."; // quick and easy 
echo "I see " . $bn->tnum( $val, ['trees', 'a tree'] ) . " on the hill."; // singular distinction
```

Output looks like this (Deutsch, English):

```html
Ich sehe neun Bäume auf dem Hügel.   (=9)
Ich sehe einen Baum auf dem Hügel.   (=1)
Ich sehe 120.000 Bäume auf dem Hügel.   (=122823) [Genauigkeit: 2 Stellen]
``` 
```html
I see nine trees on the hill.   (=9)
I see a tree on the hill.   (=1)
I see 120,000 trees on the hill.   (=122823) [accuracy: 2 digits]
```

*Ann.: We use an array for the language element so that it is easier to employ in multi-language websites, e.g. tnum($val, $LANG['de']['termin-AKK']) for the German accusative form ["Termine", "einen Termin"] and tnum($val, $LANG['de']['termin-NOM']) for the nominative ["Termine", "ein Termin"].*

## 2.2. tsyn() – text syntax ##

If you want the perfect use of numbers in running text, you might additionally need tsyn() to distinguish between singular and plural for the correlated verb (e.g. "stand" vs. "stands"). 

```php
echo $bn->tnum( $val, ['trees', 'one tree'], ['transform'=>'ucfirst'] ) . " " // start uppercase  
    . $bn->tsyn( $val, ['stand', 'stands'] ) // corresponding syntax
    . " in the market square.";

```

Output looks like this (Deutsch, English):

```html
Ein Baum steht auf dem Marktplatz.   (automatische Großschreibung)
Zwei Bäume stehen auf dem Marktplatz.
15 Bäume stehen auf dem Marktplatz.
```
```html
One tree stands in the market square.   (automatic upper case)
Two trees stand in the market square.
15 trees stand in the market square.
``` 

## 3. sinum() – statistical usage ##

For statistical usage you may want to use sinum() with a margin of error (or a tolerance) instead of an accuracy:

```php
echo $bn->sinum( 1.7781990197386, 'g', ['err'=>0.0019757766885984] ); // error instead of accuracy
```

Output looks like this (Deutsch, English):

```html
1778 ± 2 mg  (= 1.7781990197386 ± 0.0019757766885984 g)
58,97 ± 0,05 ns  (= 5.8969191908356E-8 ± 4.9140993256964E-11 s)
```
```html
1778 ± 2 mg  (= 1.7781990197386 ± 0.0019757766885984 g)
58.97 ± 0.05 ns  (= 5.8969191908356E-8 ± 4.9140993256964E-11 s)
```


Have fun!


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file (tags: Added, Changed, Deprecated, Removed, Fixed, Security).

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## 1.0.23 - 2019-05-24
- Added: Overline an ambiguous significant zero in HTML in sinum() 
- Added "circa" to language list

## 1.0.22 - 2019-05-22
- Changed: Speed improvement in sinum() 

## 1.0.21 - 2019-05-21
- Added error option 'err' to sinum() which replaces 'acc' if wanted: ['err'=>0.02]
- Added scientific number format en-SC and de-WI 

## 1.0.20 - 2019-05-17
- Changed default unit value in sinum() to empty

## 1.0.19 - 2019-05-17
- Added subsets for local language format array (accessable via "en-SI", instead of the standard "en")

## 1.0.18 - 2019-05-14
- Changed: Faster rounding 

## 1.0.17 - 2019-05-13
- Added Spanish language ("es")
- Fixed error in tnum() within 12<$val<100 (outputted "19.0" instead of "19")

## 1.0.15 - 2019-05-13
- Added rounding of big numbers ($val>100) in tnum() (120.000 instead of 119.289)

## 1.0.14 - 2019-05-13
- Changed parameters in tnum() and tsyn() to arrays (for easier use with multi-language sites)



output.php
<?php

#
# --- SirDagen PHP format functions
# --- V1.0.1 
#


// si and binary prefixes (binary prefixes below 0 make no sense, but are listed for consistency sake)
$GLOBALS['var_siprefix']=array( // decimal and binary
    -5=>array('f', 'fi'), 
    -4=>array('p', 'pi'), 
    -3=>array('n', 'ni'), 
    -2=>array('µ', 'µi'),
    -1=>array('m', 'mi'), 
    0=>array('', ''),
    1=>array('k', 'Ki'), // kilo 
    2=>array('M', 'Mi'), // mega
    3=>array('G', 'Gi'), // giga
    4=>array('T', 'Ti'),
    5=>array('P', 'Pi'), 
    ); 
    
function out_val($val, $pdp=0, $md=array()) { // post decimal places / -1 = all
    // outputs values in local number format with given decimal places
    $t='txt'; if (isset($md[$t]) and !empty($md[$t])) $$t=true; else $$t=false; // dont use HTML (e.g. &ndash;)
	if ($val===false) {
		if ($txt) return '-';
		return '&ndash;';
	}
	if ($pdp==-1) {
		$i=0; $q=$val;
		while ($q!=floor($q)) { $q*=10; $i++; }
		$pdp=$i; 
	}
	return number_format($val, $pdp, ',', '.'); // e.g. German number format
}

function out_sinum($val, $unit='B', $md=array()) { 
    // outs values with si: "3240g" -> "3.24 kg"
    $t='txt'; if (isset($md[$t]) and !empty($md[$t])) $$t=true; else $$t=false; // dont use HTML in output (e.g. &ndash;)
    $t='bin'; if (isset($md[$t]) and !empty($md[$t])) $$t=true; else $$t=false; // use IEC binary (1024) instead of SI (1000)
    $t='acc'; if (isset($md[$t])) $$t=$md[$t]; else $$t=3; // accuracy (decimal places) - preset = 3
	if ($bin===true) { $a=1024; $bin=1; }
	else { $a=1000; $bin=0; }
	$i=0;
	if (!empty($val)) {
        while (abs($val)<1) { $val*=$a; $i--; }
        while (abs($val)>$a) { $val/=$a; $i++; }
	}
	$acc-=strlen(floor(abs($val))); if ($acc<0) $acc=0; 
	if ($txt) $t=' '; else $t='&#8239;'; // &thinsp;
	$t.=$GLOBALS['var_siprefix'][$i][$bin]; 
	$rt=out_val($val, $acc, $md).$t.$unit;
	return $rt;
}




?>


demo-si.php
<?php

require('PHPzformat.php'); 

function out_demo($zformat) {
    $val=mt_rand()/mt_getrandmax(); $u='s'; 
    echo "<p>".$zformat->sinum($val, $u)." &nbsp;(= {$val} {$u})</p>"; 
    
    $val=mt_rand()/mt_getrandmax()/1000; $u='s'; 
    echo "<p>".$zformat->sinum($val, $u)." &nbsp;(= {$val} {$u})</p>";
    
    $val=mt_rand()/mt_getrandmax()/10000000; $u='s'; 
    echo "<p>".$zformat->sinum($val, $u)." &nbsp;(= {$val} {$u})</p>";
    
    $val=mt_rand()/mt_getrandmax()*2000; $u='g'; 
    echo "<p>".$zformat->sinum($val, $u)." &nbsp;(= {$val} {$u})</p>";
    
    $val=mt_rand(5, 1999999); $u='B'; 
    echo "<p>".$zformat->sinum($val, $u, ['bin'=>true])." &nbsp;(= {$val} {$u})</p>";
    
    $val=mt_rand()/mt_getrandmax()*9000; $u='m'; 
    echo "<p>".$zformat->sinum($val, $u, ['acc'=>2])." &nbsp;(= {$val} {$u})</p>"; // accuracy is set to 2 decimal digits
}


#
# --- Demo sinum: sinum($value, $unit, ['acc'=>3, 'bin'=>false])
#


// Create a new instance - German number format
$zformat = new PHPzformat\zformat([',', '.']); 
echo "<br/><p><b>Das Nummerformat ist Deutsch:</b><br/></p>"; 
out_demo($zformat);


// Create a new instance - English number format
$zformat = new PHPzformat\zformat(['.', ','], ['txt'=>false, 'acc'=>3]);
echo "<br/><p><b>The number format is set to English:</b><br/></p>"; 
out_demo($zformat);



?>


demo_de.php
<?php

require('bnformat-class.php'); 

#
# --- Demo: Deutsch
#


// Neue Instanz anlegen - Deutsches Nummberformat
$bn = new bnformat\bnformat( ['lang'=>'de'] ); 

// sinum()

echo "<br/><h3><pre>1. sinum() &ndash; Gibt Zahlen im einfach zu lesenden SI-Format aus</pre></h3>";

$val=mt_rand()/mt_getrandmax()/5; $u='s'; 
echo "<p>".$bn->sinum($val, $u)." &nbsp;(= {$val} {$u})</p>"; 

$val=mt_rand()/mt_getrandmax()/2000; $u='s'; 
echo "<p>".$bn->sinum($val, $u)." &nbsp;(= {$val} {$u})</p>";

$val=mt_rand()/mt_getrandmax()/10000000; $u='s'; 
echo "<p>".$bn->sinum($val, $u)." &nbsp;(= {$val} {$u})</p>";

$val=mt_rand()/mt_getrandmax()*2000; $u='g'; 
echo "<p>".$bn->sinum($val, $u)." &nbsp;(= {$val} {$u})</p>";

$val=mt_rand(5, 1999999); $u='B'; 
echo "<p>".$bn->sinum($val, $u, ['bin'=>true])." &nbsp;(= {$val} {$u})</p>";

$val=mt_rand()/mt_getrandmax()*9000; $u='m'; 
echo "<p>".$bn->sinum($val, $u, ['acc'=>2])." &nbsp;(= {$val} {$u}) [Genauigkeit: 2 Stellen]</p>"; // Genauigkeit = 2 Stellen

echo "<p><br/><b>kilo, Mega, Giga, Tera, ...</b></p>";

$val=round(mt_rand()/mt_getrandmax()*1000000); 
echo "<p>".$bn->sinum($val)." &nbsp;(= {$val})</p>"; 

$val=round(mt_rand()/mt_getrandmax()*90000000); 
echo "<p>".$bn->sinum($val)." &nbsp;(= {$val})</p>"; 

$val=round(mt_rand()/mt_getrandmax()*80000000000+1000000000);  
echo "<p>".$bn->sinum($val)." &nbsp;(= {$val})</p>"; 


// tnum()

echo "<br/><h3><pre>2.1. tnum() &ndash; Gibt Zahlen im Fließtext aus (0..12 werden ausgeschrieben) - text number</pre></h3>";

// Zahlen 0..12 werden ausgeschrieben, alle anderen als Zahlen ausgegeben (wie es in Publikationen üblich ist)
echo "<p>"; 
for ($i=0;$i<3;$i++) {
    $val=mt_rand(2, 15); 
    echo "Es sind ".$bn->tnum($val)." Bäume auf der Anhöhe.<br/>"; 
}
echo "</p>"; 


// tnum() - volle Nutzung

echo "<p>"; 
for ($i=0;$i<3;$i++) {
    $val=mt_rand(0, 3); 
    echo $bn->tnum($val, ['Bäume', 'einen Baum'], ['transform'=>'ucfirst'])." sehe ich am Wegesrand stehen.<br/>"; 
}
echo "</p>"; 

echo "<p><b>Und rundet auf angegebene Genauigkeit (2 Stellen):</b></p>";

echo "<p>"; 
for ($i=0;$i<3;$i++) {
    $val=mt_rand(100, 2000000); 
    echo "ca. ".$bn->tnum($val, ['Haushalte', 'ein Haushalt'])." &nbsp;(= {$val} Haushalte)<br/>"; 
}
echo "</p>"; 

/*
echo "<p>"; 
for ($i=0;$i<3;$i++) {
    $val=mt_rand(1, 2000000)/90000000;  
    echo "ca. ".$bn->tnum($val, ['Haushalte', 'ein Haushalt'])." &nbsp;(= {$val} Haushalte)<br/>"; 
}
echo "</p>"; 
*/


// tsyn() - zusätzliche Nutzung von 

echo "<br/><h3><pre>2.2. tsyn() &ndash; Unterscheidet zwischen Singular und Plural - text syntax</pre></h3>";

echo "<p>"; 
for ($i=0;$i<3;$i++) { 
    $val=mt_rand(1, 3); 
    echo "Es ".$bn->tsyn($val, ['stehen', 'steht'])." ".$bn->tnum($val, ['Bäume', 'ein Baum'])." auf dem Marktplatz.<br/>"; 
}
echo "</p>"; 


// sinum() - statistische Nutzung

echo "<br/><h3><pre>3. sinum() &ndash; Statistische Nutzung</pre></h3>";

$bnW = new bnformat\bnformat( ['lang'=>'de-WI'] ); 

$val=mt_rand()/mt_getrandmax()*2; $u='g'; $err=$val/900; 
echo "<p>".$bnW->sinum($val, $u, ['err'=>$err])." &nbsp;(= {$val} ± {$err} {$u})</p>";

$val=mt_rand()/mt_getrandmax()/10000000; $u='s'; $err=$val/1200;  
echo "<p>".$bnW->sinum($val, $u, ['err'=>$err])." &nbsp;(= {$val} ± {$err} {$u})</p>";

$val=mt_rand()/mt_getrandmax()/100; $u='m'; $err=$val/800;  
echo "<p>".$bnW->sinum($val, $u, ['err'=>$err])." &nbsp;(= {$val} ± {$err} {$u})</p>";



echo "<br/>&nbsp;<br/>";


?>


bformat-class.php
<?php

require('bformat-class.php'); 

#
# --- Demo: Deutsch
#


// Neue Instanz anlegen - Deutsches Nummberformat
$bf = new PHPbf\bformat( ['lang'=>'de'] ); 

echo "<br/><h1>php-beautiful-numbers <b>(Format = ".$bf->langname().")</b><br/></h1>"; 


// sinum()
echo "<br/><h3><pre>A. \"sinum\": Gibt Zahlen im einfach zu lesenden SI-Format aus</pre></h3>";

$val=mt_rand()/mt_getrandmax(); $u='s'; 
echo "<p>".$bf->sinum($val, $u)." &nbsp;(= {$val} {$u})</p>"; 

$val=mt_rand()/mt_getrandmax()/1000; $u='s'; 
echo "<p>".$bf->sinum($val, $u)." &nbsp;(= {$val} {$u})</p>";

$val=mt_rand()/mt_getrandmax()/10000000; $u='s'; 
echo "<p>".$bf->sinum($val, $u)." &nbsp;(= {$val} {$u})</p>";

$val=mt_rand()/mt_getrandmax()*2000; $u='g'; 
echo "<p>".$bf->sinum($val, $u)." &nbsp;(= {$val} {$u})</p>";

$val=mt_rand(5, 1999999); $u='B'; 
echo "<p>".$bf->sinum($val, $u, ['bin'=>true])." &nbsp;(= {$val} {$u})</p>";

$val=mt_rand()/mt_getrandmax()*9000; $u='m'; 
echo "<p>".$bf->sinum($val, $u, ['acc'=>2])." &nbsp;(= {$val} {$u})</p>"; // Genauigkeit = 2 Stellen


// tnum()
echo "<br/><h3><pre>B. \"tnum\": Gibt Zahlen im Fließtext aus (1..12 werden ausgeschrieben)</pre></h3>";

// Zahlen 1..12 werden ausgeschrieben, alle anderen als Zahlen ausgegeben (wie es in Publikationen üblich ist)
echo "<p>"; 
for ($i=0;$i<3;$i++) {
    $val=mt_rand(2, 15); 
    echo "Es sind ".$bf->tnum($val)." Bäume auf der Anhöhe.<br/>"; 
}
echo "</p>"; 


echo "<p><i>Aber auch Kommazahlen (wenn benötigt):</i></p>";

$val=mt_rand()/mt_getrandmax()*2;
echo "<p>Das Mittel liegt bei ".$bf->tnum($val, 1)." Kindern.</p>"; 



echo "<br/>&nbsp;<br/>";


?>


bnformat-class.php
<?php
/*
 * bnformat     php-beautiful-numbers class.
 * Version      -see below-
 *
 * @see         https://github.com/SirDagen/php-beautiful-numbers
 *
 * @author      Gordon Axmann
 * @note        This program is licensed under the GNU General Public License v3.0
 *              Copyright and license notices must be preserved. See details at:
 *              https://github.com/SirDagen/php-beautiful-numbers/blob/master/LICENSE 
 */

namespace bnformat;

/*
 * Name         php-beautiful-numbers class (number format functions)
 * Version      1.0.23
 * @author      Gordon Axmann
 */

class bnformat {

    /*

    --- INIT

    $bn = new bnformat\bnformat([ 'lang'=>'en', 'txt'=>false, 'acc'=>3, 'numberformat'=> ['.', ','] ]); // (if you state the language, the number format will be set automatically accordingly) 
    $bn = new bnformat\bnformat([ 'lang'=>'en-SI' ]); // you can also set a sub type of the local language format


    --- QUICK MANUAL

    $bn->sinum( 7833.6227931239, 'm', ['acc'=>2] ); // works with multiple languages, accuracy, number formats, ... // = 7.8 km (English) -OR- 7,8 km (German)
    $bn->sinum( 0.00050260131503576, 's' ); // outputs numbers in easy to read SI prefix format // = 503 µs
    $bn->sinum( 404436, 'B', ['bin'=>true] ); // sinum() also works with the binary system // = 395 KiB
    
    $bn->tnum( 9 ); // outputs numbers for running text (0..12 will be written-out) // = nine
    $bn->tnum( 5, ['Bäume', 'einem Baum'] ); // you can also make tnum() choose between the plural and the singular word
    
    $bn->tsyn( 5, ['stehen', 'steht'] ); // for the perfect use of numbers in running text you might have to use tsyn() to select the corresponding syntax of the verb (see demo file)
    
    */

    // all presets can be overwriten when using the constructor // $bn = new bnformat\bnformat([ 'lang'=>'en' ); 
    var $presets=['lang'=>'de', 'txt'=>false, 'acc'=>3, 'over'=>true, 'numberformat'=> [',', '.'] ]; 
    // (lang)uage - language code: de, en, fr, de-WI, ... 
    // (txt) - Text(=true) or HTML(=false) 
    // (acc)uracy - significant figures https://en.wikipedia.org/wiki/Significant_figures
    // (over)line - overline ambiguous significant zero in HTML
    // (numberformat) - set number format (usually this gets set automatically by choosing the correct language code)

    // SI and binary prefixes (binary prefixes below 0 make no sense '?')
    // https://en.wikipedia.org/wiki/International_System_of_Units
    var $siprefix=array( // SI decimal, IEC binary
        -8=>['y', '?'], // yocto
        -7=>['z', '?'], // zepto
        -6=>['a', '?'], // atto
        -5=>['f', '?'], // femto
        -4=>['p', '?'], // pico
        -3=>['n', '?'], // nano
        -2=>['µ', '?'], // micro
        -1=>['m', '?'], // milli
         0=>['', ''],
         1=>['k', 'Ki'], // kilo, ("Ki") <- big K, sic(!)
         2=>['M', 'Mi'], // mega
         3=>['G', 'Gi'], // giga
         4=>['T', 'Ti'], // tera
         5=>['P', 'Pi'], // peta
         6=>['E', 'Ei'], // exa
         7=>['Z', 'Zi'], // zetta
         8=>['Y', 'Yi'], // yotta 
        ); 

        
    
    // numbers 0..12 are usually written-out in publications/running text 
    // the language of the constructor (see INIT) also sets the local number format
    // [ 'lang'=>'de' ] 
    // or by using the subvariants of the language arrays, like "de-CH", which will overwrite the standard 'numberformat' entry set by 'lang'
    // [ 'lang'=>'de-CH' ] // set language to de, language format to the subset "de-CH"
    var $numwords=array(
        'de'=> [ 'null', 'ein/e/m', 'zwei', 'drei', 'vier', 'fünf', 'sechs', 'sieben', 'acht', 'neun',
                'zehn', 'elf', 'zwölf', 'minusword'=>'minus', 'langname'=>'Deutsch', 'circa'=>'ca.',
                // std number format for this language
                'numberformat'=> [',', '.'], // dec_point, thousands_sep 
            ],
        'en'=> [ 'zero', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine',
                'ten', 'eleven', 'twelve', 'minusword'=>'minus', 'langname'=>'English', 'circa'=>'ca.',
                // std number format for this language
                'numberformat'=> ['.', ','], // dec_point, thousands_sep 
            ],
        'fr'=> [ 'zéro', 'un/une', 'deux', 'trois', 'quatre', 'cinq', 'six', 'sept', 'huit', 'neuf',
                'dix', 'onze', 'douze', 'minusword'=>'moins', 'langname'=>'français', 'circa'=>'environ',
                // std number format for this language
                'numberformat'=> [',', ' '], // dec_point, thousands_sep 
            ],
        'es'=> [ 'cero', 'uno', 'dos', 'tres', 'cuatro', 'cinco', 'seis', 'siete', 'ocho', 'nueve',
                'diez', 'once', 'doce', 'minusword'=>'menos', 'langname'=>'español', 'circa'=>'aprox.', 
                // std number format for this language
                'numberformat'=> [',', '.'], // dec_point, thousands_sep 
        ],
    );

    // local sub number formats that differ from the above language array
    // can be accessed by using the extended language code, e.g. "en-SI"
    // if you need yet another number format you can individually specify it in the __constructor via: 
    // [ 'lang'=>'en', 'numberformat'=> ["'", '.'] ] 
    // local number format: https://en.wikipedia.org/wiki/Decimal_separator#Examples_of_use
    var $numformat=array(
        // type => [ dec_point, thousands_sep ]
        'en-SI'=> ['.', ' '], // English-SI style (official English-Canada) 
        'en-MT'=> ['·', ' '], // English-Malta
        'en-SC'=> ['.', ''], // English-scientific 
        'de-LI'=> ['.', "'"], // Deutsch-Liechtenstein  
        'de-WI'=> [',', ''], // Deutsch-Wissenschaftlich 
    );
    
    function __construct($presets=false) { // $presets=array() -- INIT
        // overwrite the class presets (if wanted), e.g. $presets = [ 'lang'=>'en', 'numberformat'=> ['.', ' '], 'acc'=>4 ]
        $numformat=false;
        if (is_array($presets)) {
            if (isset($presets['lang']) and (strpos($presets['lang'], '-')!==false)) { 
                if (array_key_exists($presets['lang'], $this->numformat)) $numformat=$this->numformat[$presets['lang']];
                $lgar=explode('-', $presets['lang']);
                $presets['lang']=$lgar[0]; // only take first part of 'lang', e.g. "en" instead "en-SI" 
            }
            foreach ($presets as $k0=>$v0) $this->presets[$k0] = $v0; 
        }
        // when numberformat is not explicitly stated, get local numberformat from presets['lang']   
        if (!is_array($presets['numberformat'])) {
            if (!empty($numformat)) $this->presets['numberformat']=$numformat; // use a number format from the $numformat[] subset, e.g. "en-SI"
            else $this->presets['numberformat']=$this->numwords[$this->presets['lang']]['numberformat']; // use number format from $numwords[], e.g. "de" 
        }
    }


    function langname() {
        return $this->numwords[$this->presets['lang']]['langname']; // outputs: "Deutsch" 
    }


    function out_val($val, $pdp=0, $md=[]) { // post decimal places (99 = all)
        // outputs numbers in local number format (with stated decimal places) 
        // it is a copy of number_format but with better rounding
        $t='txt'; if (isset($md[$t])) $$t=$md[$t]; else $$t=$this->presets[$t]; // !dont use HTML entities in output (e.g. &ndash;)
        $t='over'; if (isset($md[$t])) $$t=$md[$t]; else $$t=$this->presets[$t]; 
        if ($val===false) {
            if ($txt) return '-';
            return '&ndash;';
        }
        if ($pdp==99) {
            $pdp=0; $q=$val;
            while ($q!=floor($q)) { $q*=10; $pdp++; }
        }
        if ($pdp<0) { // force rounding when there are more digits before the decimal point than the given accuracy
            $val=round($val, $pdp); 
            // $base=pow(10, -(int)$pdp); $val=round($val/$base)*$base; // older version 
        }
        $rt=number_format($val, $pdp, $this->presets['numberformat'][0], $this->presets['numberformat'][1]); 
        // if a ambiguous significant zero exists, mark it in HTML with overline
        // https://en.wikipedia.org/wiki/Significant_figures#Identifying_significant_figures
        if ($over and !$txt and ($pdp<=0)) {
            $i=strlen($rt)-1; $p=0;
            while ($p>$pdp) { 
                if (is_numeric(substr($rt, $i, 1))) $p--;
                $i--; 
            } 
            if (substr($rt, $i, 1)=='0') $rt=substr($rt, 0, $i)."<span style='text-decoration:overline;'>0</span>".substr($rt, $i+1); // &#773; is ugly
        }
        return $rt;
    }
 

    function sinum($val, $unit='', $md=[]) { // SI number output
        // outs values with SI: "3240g" -> "3.24 kg"
        // https://en.wikipedia.org/wiki/International_System_of_Units
        $t='txt'; if (isset($md[$t])) $$t=$md[$t]; else $$t=$this->presets[$t]; // !dont use HTML entities in output (e.g. &ndash;)
        $t='acc'; if (isset($md[$t])) $$t=$md[$t]; else $$t=$this->presets[$t]; // accuracy (= decimal digits) - preset = 3
        $t='err'; if (isset($md[$t])) { $$t=$md[$t]; if (!isset($md['over'])) $md['over']=false; } else $$t=0; // use (err)or value instead of (acc)uracy
        $t='bin'; if (isset($md[$t]) and !empty($md[$t])) $$t=$md[$t]; else $$t=false; // use IEC binary (1024) instead of SI (1000)
        if ($bin===true) { $base=1024; $ptype=1; } else { $base=1000; $ptype=0; }
        // $pow=floor(log($val, $base)); $val/=pow($base, $pow); // dont know what way is faster in average
        $pow=0;
        if (!empty($val)) {
            while (abs($val)<1) { $val*=$base; $err*=$base; $pow--; }
            while (abs($val)>1000) { $val/=$base; $err/=$base; $pow++; } // 0,984 MiB is easier to read than 1.010 KiB  
            if (!empty($err)) while (abs($err)<0.0095) { $val*=$base; $err*=$base; $pow--; } // this(!) order of whiles(!)
        }
        $acc-=strlen(floor(abs($val))); //if ($acc<0) $acc=0; // only positive values supported right now
        if (floor(abs($val))==0) $acc++; // add a digit if integer portion is zero (e.g. 0.984 MiB)
        $prefix=$this->siprefix[$pow][$ptype];
        if (empty($unit) and empty($prefix)) $sp='';
        else {
            if ($txt) $sp=' '; else $sp='&#8239;'; // = &thinsp;
        }
        if (empty($err)) $rt=$this->out_val($val, $acc, $md).$sp.$prefix.$unit;
        else {
            $lg=log(abs($err), 10); $frac=($lg<0) ? $lg-ceil($lg) : $lg-floor($lg); // $frac=fmod($lg, 1); is a little slower
            $acc=-floor($lg); 
            if ($frac<0 && $frac>-0.02227639471121) $acc--; // dont display error = 1.0 (substract 1 digit of accuracy, if e.g. 0.99827199496456)
            $rt=$this->out_val($val, $acc, $md).$sp.'±'.$sp;
            $t='over'; $md[$t]=$this->presets[$t]; // sets 'over' to presets
            $rt.=$this->out_val($err, $acc, $md).$sp.$prefix.$unit; 
        }
        return $rt;
    }

    
    function ucfirst($tx) { // upper case first letter
        if (empty($tx)) return $tx;
        return mb_strtoupper(mb_substr($tx, 0, 1)).mb_substr($tx, 1);
    }
            
    function lcfirst($tx) { // lower case first letter
        if (empty($tx)) return $tx;
        return mb_strtolower(mb_substr($tx, 0, 1)).mb_substr($tx, 1); 
    }
    
        
    function tnum($val, $syntax=false, $md=[]) {  // text number // $syntax = array('plural', 'full_singular')
        // writes integer numbers 0..12 written-out. all others as round digits (for running text)
        // you can use it to distinguish between singular and plural. PLEASE NOTE, that you have to offer the FULL SINGULAR, e.g. "one tree" or "a tree" (!) 
        // the standard accuracy for tnum is 2
        if (($syntax!==false) and !is_array($syntax)) echo "***?ERROR-1 tnum()*** ";
        if (is_array($syntax) and !array_key_exists(0, $syntax)) echo "***?ERROR-2 tnum()*** ";
        $t='acc'; if (isset($md[$t])) $$t=$md[$t]; else $$t=2; // accuracy (= decimal digits) 
        $t='pdp'; if (isset($md[$t])) $$t=$md[$t]; else $$t=0; // post decimal places (99 = all) 
        $t='lang'; if (isset($md[$t])) $$t=$md[$t]; else $$t=$this->presets[$t]; // choose language
        $t='transform'; if (!isset($md[$t])) $$t=false; else $$t=$md[$t]; // apply (ucfirst OR toupper) to written-out number
        $base=1000; 
        // >12 or fractional  
        if (($pdp!=0) or (abs($val)>12) or ((abs($val)<0.995) and (abs($val)>0))) {
            if ($pdp!=0) $rt=$this->out_val($val, $pdp, $md);
            else {
                $pow=0; $a=$val;
                if (!empty($val)) {
                    while (abs($a)<1) { $a*=$base; $pow--; }
                }
                //$pow2=floor(log($val, 10)); // doesnt work
                $acc-=strlen(floor(abs($a)))+3*$pow; //if ($acc<0) $acc=0; // only positive values supported right now
                if (($acc>0) and (abs($val)>=10)) $acc=0; // dont show e.g. "15.0 trees"
                $rt=$this->out_val($val, $acc, $md);
            }
            if (is_array($syntax)) $rt.=' '.$syntax[0]; 
        }
        // 0..12
        else {
            $val=round($val); 
            // number as word
            $rt=$this->numwords[$lang][abs($val)];
            // add object text
            if (is_array($syntax)) {
                if (abs($val)==1) $rt=$syntax[1];
                else $rt.=' '.$syntax[0];
            }
            if ($val<0) {
                $rt=$this->numwords[$lang]['minusword'].' '.$rt;
                $val=-$val; 
                }
            if (!empty($transform)) {
                switch($transform) {
                    case 'ucfirst':
                    $rt=$this->ucfirst($rt);
                    break;

                    case 'upper': 
                    $rt=mb_strtoupper($rt);
                    break;
                }
            }
        }
        return $rt;
    }

function tsyn($val, $syntax, $md=[]) { // text syntax // $syntax = array('plural', 'singular')
    // chooses between the use of plural or singular
    if (!is_array($syntax)) echo "***?ERROR-1 tsyn()*** ";
    if (is_array($syntax) and !array_key_exists(0, $syntax)) echo "***?ERROR-2 tsyn()*** ";
    $t='transform'; if (!isset($md[$t])) $$t=false; else $$t=$md[$t]; // apply (ucfirst OR toupper) to written-out number
    $val=round($val); 
    if (abs($val)==1) $rt=$syntax[1];
    else $rt=$syntax[0];
    if (!empty($transform)) {
        switch($transform) {
            case 'ucfirst':
            $rt=$this->ucfirst($rt);
            break;

            case 'upper':
            $rt=mb_strtoupper($rt);
            break;
        }
    }
    return $rt; 
}



} // end of class

?>
